backend/pkg/api/connect/service/security/service.go

```diff
@@ -0,0 +1,332 @@
+		return nil, apierrors.NewRedpandaAdminAPINotConfiguredError()
+	}
+
+	// Check if logged in user is allowed to list roles for the given request
+	canListRoles, restErr := s.authHooks.CanListKafkaRoles(ctx)
+	if restErr != nil || !canListRoles {
+		if restErr != nil {
+			err := errors.New("you don't have permissions to view Kafka roles")
+			if restErr.Err != nil {
+				err = restErr.Err
+			}
+			return nil, apierrors.NewConnectError(
+				connect.CodePermissionDenied,
+				err,
+				apierrors.NewErrorInfo(commonv1alpha1.Reason_REASON_PERMISSION_DENIED.String()),
+			)
+		}
+	}
+
+	res, err := s.redpandaSvc.RoleMembers(ctx, req.Msg.GetRoleName())
+	if err != nil {
+		return nil, apierrors.NewConnectErrorFromRedpandaAdminAPIError(err, "")
+	}
+
+	nameContains := req.Msg.GetFilter().GetNameContains()
+	var out []adminapi.RoleMember
+	if nameContains == "" {
+		out = res.Members
+	} else {
+		out = make([]adminapi.RoleMember, 0, len(res.Members))
+		for _, m := range res.Members {
+			if strings.Contains(m.Name, nameContains) {
+				out = append(out, m)
+			}
+		}
+	}
+
+	protoMembers := adminAPIRoleMembersToProto(out)
+
+	// Add pagination
+	var nextPageToken string
+	if req.Msg.GetPageSize() > 0 {
+		sort.SliceStable(protoMembers, func(i, j int) bool {
+			return protoMembers[i].Principal < protoMembers[j].Principal
+		})
+
+		page, token, err := pagination.SliceToPaginatedWithToken(protoMembers, int(req.Msg.PageSize), req.Msg.GetPageToken(), "principal", func(x *v1alpha1.RoleMembership) string {
+			return x.GetPrincipal()
+		})
+		if err != nil {
+			return nil, apierrors.NewConnectError(
+				connect.CodeInternal,
+				fmt.Errorf("failed to apply pagination: %w", err),
+				apierrors.NewErrorInfo(v1alpha1dp.Reason_REASON_CONSOLE_ERROR.String()),
+			)
+		}
+		nextPageToken = token
+		protoMembers = page
+	}
+
+	return connect.NewResponse(&v1alpha1.ListRoleMembersResponse{
+		RoleName:      req.Msg.GetRoleName(),
+		Members:       protoMembers,
+		NextPageToken: nextPageToken,
+	}), nil
+}
+
+// UpdateRoleMembership updates role membership, adding and removing principals assigned to the role.
+func (s *Service) UpdateRoleMembership(ctx context.Context, req *connect.Request[v1alpha1.UpdateRoleMembershipRequest]) (*connect.Response[v1alpha1.UpdateRoleMembershipResponse], error) {
+	if !s.cfg.Redpanda.AdminAPI.Enabled {
+		return nil, apierrors.NewRedpandaAdminAPINotConfiguredError()
+	}
+
+	// Check if logged in user is allowed to create roles for the given request
+	canCreateRoles, restErr := s.authHooks.CanCreateKafkaRoles(ctx)
+	if restErr != nil || !canCreateRoles {
+		if restErr != nil {
+			err := errors.New("you don't have permissions to create Kafka roles")
+			if restErr.Err != nil {
+				err = restErr.Err
+			}
+			return nil, apierrors.NewConnectError(
+				connect.CodePermissionDenied,
+				err,
+				apierrors.NewErrorInfo(commonv1alpha1.Reason_REASON_PERMISSION_DENIED.String()),
+			)
+		}
+	}
```

Comment:
This is very hard to read and more nested than necessary. In fact it's not handling the case where restErr is nil & canCreateRoles = false. 

This would be much quicker to digest (even though a bit more verbose). FWIW the `restErr` will only be set if there's an issue such as the user session not being set in the request context. It's more like an internal server error, but we can still reject the request as permission denied in that case.

```go
	// Check if logged in user is allowed to create roles for the given request
	canCreateRoles, restErr := s.authHooks.CanCreateKafkaRoles(ctx)
	if restErr != nil {
	      return nil, apierrors.NewConnectError(
		      connect.CodePermissionDenied,
		      // The returned `restErr` is in our control, so whether or not Err is set
		      // should be in our control. It should always be set!
		      restErr.Err,
		      apierrors.NewErrorInfo(commonv1alpha1.Reason_REASON_PERMISSION_DENIED.String()),
	      )
	}

	if !canCreateRoles {
	    return nil, apierrors.NewConnectError(
		      connect.CodePermissionDenied,
		      errors.New("You are not allowed to create roles")
		      apierrors.NewErrorInfo(commonv1alpha1.Reason_REASON_PERMISSION_DENIED.String()),
	      )
	}
```